timers/timers_test.go

```diff
@@ -0,0 +1,417 @@
+		assert.True(t, canceled.Stop())
+	}
+
+	for elapsed := time.Duration(0); elapsed < maxDelay; elapsed += tickDuration {
+		wg.Add(1)
+		c.Add(tickDuration)
+		wg.Wait()
+	}
+
+	for i := range scheduled {
+		assert.False(t, scheduled[i].Stop(), "Shouldn't be able to cancel after expiry.")
+	}
+}
+
+func TestTimerDroppingTicks(t *testing.T) {
+	const (
+		numTimers    = 100
+		tickDuration = 1 << 22
+		maxDelay     = numTimers * tickDuration
+	)
+	w, c := fakeWheel(tickDuration)
+	defer w.Stop()
+
+	var wg sync.WaitGroup
+	wg.Add(numTimers)
+	timeouts := randomTimeouts(numTimers, maxDelay)
+	for _, d := range timeouts {
+		w.AfterFunc(d, func() { wg.Done() })
+	}
+	c.Add(maxDelay * 2)
+	wg.Wait()
+}
+
+func TestTimerClearAfterStop(t *testing.T) {
+	const numTimers = 100
+	w, _ := fakeWheel(1 * time.Millisecond)
+
+	var wg sync.WaitGroup
+	wg.Add(numTimers)
+	timeouts := randomTimeouts(numTimers, 100*time.Millisecond)
+	for _, d := range timeouts {
+		w.AfterFunc(d, func() { wg.Done() })
+	}
+
+	w.Stop()
+	wg.Wait()
+}
+
+func TestPowersOfTwo(t *testing.T) {
+	tests := []struct {
+		in       int
+		out      int
+		exponent int
+	}{
+		{-42, 1, 0},
+		{0, 1, 0},
+		{1, 1, 0},
+		{5, 8, 3},
+		{1000, 1024, 10},
+		{1 << 22, 1 << 22, 22},
+	}
+	for _, tt := range tests {
+		n, exp := nextPowerOfTwo(int64(tt.in))
+		assert.Equal(t, tt.out, int(n), "Unexpected next power of two for input %v", tt.in)
+		assert.Equal(t, tt.exponent, int(exp), "Unexpected exponent for input %v", tt.in)
+	}
+}
+
+func scheduleAndCancel(b *testing.B, w *Wheel) {
+	ds := randomTimeouts(1024, time.Minute)
+	for i := range ds {
+		ds[i] = ds[i] + 24*time.Hour
+	}
+	b.ResetTimer()
+
+	b.RunParallel(func(pb *testing.PB) {
+		i := 0
+		for pb.Next() {
+			w.AfterFunc(ds[i&10], fakeWork).Stop()
+			i++
+		}
+	})
+}
+
+func scheduleAndCancelStd(b *testing.B) {
+	ds := randomTimeouts(1024, time.Minute)
+	for i := range ds {
+		ds[i] = ds[i] + 24*time.Hour
+	}
+	b.ResetTimer()
+
+	b.RunParallel(func(pb *testing.PB) {
+		i := 0
+		for pb.Next() {
+			time.AfterFunc(ds[i&10], fakeWork).Stop()
+			i++
+		}
+	})
+}
+
+func BenchmarkScheduleAndCancelWheelNoHeap(b *testing.B) {
+	w := NewWheel(5*time.Millisecond, 2*time.Minute)
+	defer w.Stop()
+
+	scheduleAndCancel(b, w)
+	b.StopTimer()
+}
+
+func BenchmarkScheduleAndCancelWheelWithHeap(b *testing.B) {
+	w := NewWheel(5*time.Millisecond, 2*time.Minute)
+	defer w.Stop()
+
+	ds := randomTimeouts(10000, time.Minute)
+	for i := range ds {
+		ds[i] = ds[i] + 24*time.Hour
+	}
+	for i := range ds {
+		w.AfterFunc(ds[i], fakeWork)
+	}
+
+	scheduleAndCancel(b, w)
+	b.StopTimer()
+}
+
+func BenchmarkScheduleAndCancelStandardLibraryNoHeap(b *testing.B) {
+	scheduleAndCancelStd(b)
+}
+
+func BenchmarkScheduleAndCancelStandardLibraryWithHeap(b *testing.B) {
+	ds := randomTimeouts(10000, time.Minute)
+	timers := make([]*time.Timer, 10000)
+	for i := range ds {
+		ds[i] = ds[i] + 24*time.Hour
+	}
+	for i := range ds {
```

Comment:
Is there a reason for a separate for loop rather than moving this logic into the above one?